### PR TITLE
API for no-op proxy between device and Key server

### PR DESCRIPTION
**Pull request by jendakol**
Part of #73.

### CODE REVIEW

File: pkg/api/v1/publish-api-types.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	reportType := ReportType(s)
+	switch reportType {
+	case Unknown, ConfirmedTest, ConfirmedClinicalDiagnosis, SelfReport, Recursive, Revoked:
+		*lt = reportType
+		return nil
+	}
+	return errors.New("Invalid leave type")
+}
```
Comment (marekaf) [inline pkg/api/v1/publish-api-types.go]: isn't this more like some internals/utils.go rather than public api?:)

**Review by marekaf — COMMENTED** [review]
Isn't most of this just boilerplate? We can easily import this https://github.com/google/exposure-notifications-server/blob/main/pkg/api/v1/exposure_types.go rather than copy-paste all of it.

File: pkg/api/v1/publish-api-types.go (hunk 1 above)
Comment (jendakol) [reply pkg/api/v1/publish-api-types.go]: I don't know. I was that much pissed Go lack so basic feature I didn't really think about that. But you're probably right...

**Comment by jendakol** [comment]
Good point. We can import the file and just create type aliases matching our naming scheme.

**Comment by jendakol** [comment]
@marekaf Updated.

It required running `go mod tidy` which cause quite a big change in go module metafiles.

File: pkg/api/v1/publish-api-types.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	SymptomOnsetInterval int32               `json:"symptomOnsetInterval"`
+	Traveler             bool                `json:"traveler"`
+	RevisionToken        string              `json:"revisionToken" validate:"required"`
+
+	VisitedCountries    []string   `json:"visitedCountries"`
+	ReportType          ReportType `json:"reportType"`
+	ConsentToFederation bool       `json:"consentToFederation"`
+
+	Padding string `json:"padding"`
+}
```
Comment (marekaf) [inline pkg/api/v1/publish-api-types.go]: this has the same problem as before

why not
```
type PublishKeysRequestDevice struct {
         keyserverapi.Publish //embedded struct

	VisitedCountries    []string   `json:"visitedCountries"`
	ReportType          ReportType `json:"reportType"`
	ConsentToFederation bool       `json:"consentToFederation"`
}
```

File: pkg/api/v1/publish-api-types.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+//
+// The Padding field may be populated with random data on both success and
+// error responses.
+type PublishKeysResponseDevice struct {
+	RevisionToken     string `json:"revisionToken,omitempty"`
+	InsertedExposures int    `json:"insertedExposures,omitempty"`
+	ErrorMessage      string `json:"error,omitempty"`
+	Code              string `json:"code,omitempty"`
+	Padding           string `json:"padding,omitempty"`
+}
```
Comment (marekaf) [inline pkg/api/v1/publish-api-types.go]: this is 100% equal to upstream struct

File: pkg/api/v1/publish-api-types.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+//    is not set, then a report type of
+//     CONFIRMED will lead to TR 2
+//     LIKELY will lead to TR 4
+//     NEGATIVE will lead to TR 6
+type DeviceExposureKey struct {
+	Key              string `json:"key" validate:"required"`
+	IntervalNumber   int32  `json:"rollingStartNumber" validate:"required"`
+	IntervalCount    int32  `json:"rollingPeriod" validate:"required"`
+	TransmissionRisk int    `json:"transmissionRisk"`
+}
```
Comment (marekaf) [inline pkg/api/v1/publish-api-types.go]: same things here again

File: pkg/api/v1/publish-api-types.go (hunk 3 above)
Comment (jendakol) [reply pkg/api/v1/publish-api-types.go]: Yes. At this point. Are you sure there won't be a need for extension in the future? Or will we solve it then? :confused:

File: pkg/api/v1/publish-api-types.go (hunk 2 above)
Comment (jendakol) [reply pkg/api/v1/publish-api-types.go]: Oh. Didn't know about this syntax :smile:

File: pkg/api/v1/publish-api-types.go (hunk 3 above)
Comment (marekaf) [reply pkg/api/v1/publish-api-types.go]: it's upstream v1 api, they do a really good job by no introducing breaking changes to it and handle deprecation correctly (before there was v1alpha1 which they still support). It shouldn't happen then.

File: pkg/api/v1/publish-api-types.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+//    key uploads)
+//
+// Traveler - set to true if the TEKs in this publish set are consider to be the
+//  keys of a "traveler" who has left the home region represented by this server
+//  (or by the home health authority in case of a multi-tenant installation).
+//
+// RevisionToken: An opaque string that must be passed intact on additional
+//   publish requests from the same device, where the same TEKs may be published
+//   again.
+//
```
Comment (marekaf) [inline pkg/api/v1/publish-api-types.go]: can we remove this?

File: pkg/api/v1/publish-api-types.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// VisitedCountries: list (possibly empty) where the device has travelled to.
+//
+// ReportType: type of report - is it self-report, confirmed diagnose, ...?
+//
+// ConsentToFederation: whether user of the device has allowed to share his keys to other countries.
+//
+// Padding: random base64 encoded data to obscure the request size. The server will
+// not process this data in any way.
+//
```
Comment (marekaf) [inline pkg/api/v1/publish-api-types.go]: and this

**Comment by marekaf** [comment]
much better:)

File: pkg/api/v1/publish-api-types.go (hunk 3 above)
Comment (jendakol) [reply pkg/api/v1/publish-api-types.go]: No no, I meant what happens when I need to diverge the response to a device from the standard one. But It seems improbable when I think about that.

Comment (marekaf) [reply pkg/api/v1/publish-api-types.go]: well that should be handled the same, create a new api version, like `v2alpha1` and keep the old one for backward compatiblity :)

**Review by marekaf — APPROVED** [review]
lgtm!

**Pull request merged**